File: Sean的leetcode/Leetcode/Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode *root, int sum, int &total_sum)
    {
        if (!root) return;

        sum = sum * 10 + root->val;

        if (!root->left && !root->right)
        {
            total_sum += sum;
            return;
        }
        dfs(root->left, sum, total_sum);
        dfs(root->right, sum, total_sum);
    }

    int sumNumbers(TreeNode* root) {
        int total = 0;
        dfs(root, 0, total);
        return total;
    }
};

File: MyLeetCode/python/Sum Root to Leaf Numbers.py
###
# Time Complexity: O(n) traverse n node
# Space Complexity: O(n) recursive stack
###
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        return self.dfs(root, 0)

    def dfs(self, cur, one_sol):
        if not cur:
            return 0
        if not cur.left and not cur.right:
            return one_sol * 10 + cur.val
        return self.dfs(cur.left, one_sol * 10 + cur.val) + self.dfs(cur.right, one_sol * 10 + cur.val)